In [1]:
import numpy as np

In [2]:
x = np.array([1,2,3,4,5,6,1,2,3,4,5])

In [3]:
x>=3

array([False, False,  True,  True,  True,  True, False, False,  True,
        True,  True])

In [4]:
np.where(x>=3) #indices of the trues

(array([ 2,  3,  4,  5,  8,  9, 10]),)

In [5]:
np.where(x>=3, x, 0)

array([0, 0, 3, 4, 5, 6, 0, 0, 3, 4, 5])

In [6]:
import pandas as pd
#numpy is a lower level and pandas are higher level, so pandas dfs can use all the function of numpy

In [7]:
area = [10,20,30,40]
country = ["Brazil", "Russia", "India", "China"]
capital = ["Brasilia", "Moscow", "New Delhi", "Beijing"]

In [8]:
df=pd.DataFrame({"country": country, "capital": capital, "area": area}, index = ["A","B", "C", "D"], columns = ["country", "capital", "area"])
df

country    capital  area
A  Brazil   Brasilia    10
B  Russia     Moscow    20
C   India  New Delhi    30
D   China    Beijing    40

In [9]:
#pandas dataframes are 2d heterogenous data structures
#pandas series are 1d heterogenous data structures - each row / column of a df is 1 series


In [10]:
population = [100,300,500,200]

In [11]:
#to add a column

In [12]:
df["Population"] = population

In [13]:
df

country    capital  area  Population
A  Brazil   Brasilia    10         100
B  Russia     Moscow    20         300
C   India  New Delhi    30         500
D   China    Beijing    40         200

In [14]:
s = df["country"]

In [15]:
type(s)

pandas.core.series.Series

In [16]:
c = df["capital"]

In [17]:
type(c)

pandas.core.series.Series

In [18]:
a = df["area"]

In [19]:
type(a)

pandas.core.series.Series

In [20]:
df

country    capital  area  Population
A  Brazil   Brasilia    10         100
B  Russia     Moscow    20         300
C   India  New Delhi    30         500
D   China    Beijing    40         200

In [21]:
s = pd.Series({'country': "South Africa", "capital": "Pretoria", "area": 5, "Population": 50}, name = "E")

In [22]:
s

country       South Africa
capital           Pretoria
area                     5
Population              50
Name: E, dtype: object

In [23]:
df = df.append(s, ignore_index=False)

In [24]:
df

country    capital  area  Population
A        Brazil   Brasilia    10         100
B        Russia     Moscow    20         300
C         India  New Delhi    30         500
D         China    Beijing    40         200
E  South Africa   Pretoria     5          50

#### Row indices and column names

In [25]:
df.columns

Index(['country', 'capital', 'area', 'Population'], dtype='object')

In [26]:
df.index

Index(['A', 'B', 'C', 'D', 'E'], dtype='object')

In [27]:
df.columns = ["Co", "Cap", "Ar", "Pop"]

In [28]:
df

Co        Cap  Ar  Pop
A        Brazil   Brasilia  10  100
B        Russia     Moscow  20  300
C         India  New Delhi  30  500
D         China    Beijing  40  200
E  South Africa   Pretoria   5   50

In [29]:
df.index = np.arange(5)

In [30]:
df

Co        Cap  Ar  Pop
0        Brazil   Brasilia  10  100
1        Russia     Moscow  20  300
2         India  New Delhi  30  500
3         China    Beijing  40  200
4  South Africa   Pretoria   5   50

In [31]:
df.columns[2] = "Area"

TypeError: Index does not support mutable operations

In [52]:
col=list(df.columns)

In [53]:
col[2] = "Area"
col

['Co', 'Cap', 'Area', 'Pop']

In [54]:
df.columns = col

In [55]:
df

Co        Cap  Area  Pop
0        Brazil   Brasilia    10  100
1        Russia     Moscow    20  300
2         India  New Delhi    30  500
3         China    Beijing    40  200
4  South Africa   Pretoria     5   50

In [56]:
df = df.reset_index()

In [57]:
df

index            Co        Cap  Area  Pop
0      0        Brazil   Brasilia    10  100
1      1        Russia     Moscow    20  300
2      2         India  New Delhi    30  500
3      3         China    Beijing    40  200
4      4  South Africa   Pretoria     5   50

In [58]:
df=df.drop("index", axis = 1)

In [59]:
df

Co        Cap  Area  Pop
0        Brazil   Brasilia    10  100
1        Russia     Moscow    20  300
2         India  New Delhi    30  500
3         China    Beijing    40  200
4  South Africa   Pretoria     5   50

In [60]:
df.reset_index(inplace=True)

In [61]:
df

index            Co        Cap  Area  Pop
0      0        Brazil   Brasilia    10  100
1      1        Russia     Moscow    20  300
2      2         India  New Delhi    30  500
3      3         China    Beijing    40  200
4      4  South Africa   Pretoria     5   50

In [62]:
df.drop("index", axis = 1, inplace = True)

In [63]:
df

Co        Cap  Area  Pop
0        Brazil   Brasilia    10  100
1        Russia     Moscow    20  300
2         India  New Delhi    30  500
3         China    Beijing    40  200
4  South Africa   Pretoria     5   50

In [64]:
df.set_index("Co", inplace = True)

In [65]:
df

Cap  Area  Pop
Co                                
Brazil         Brasilia    10  100
Russia           Moscow    20  300
India         New Delhi    30  500
China           Beijing    40  200
South Africa   Pretoria     5   50

In [66]:
df = df.reset_index()

In [67]:
df

Co        Cap  Area  Pop
0        Brazil   Brasilia    10  100
1        Russia     Moscow    20  300
2         India  New Delhi    30  500
3         China    Beijing    40  200
4  South Africa   Pretoria     5   50

In [68]:
df.rename(columns = {"Co": "Country", "Cap": "Capital"})

Country    Capital  Area  Pop
0        Brazil   Brasilia    10  100
1        Russia     Moscow    20  300
2         India  New Delhi    30  500
3         China    Beijing    40  200
4  South Africa   Pretoria     5   50

In [69]:
df

Co        Cap  Area  Pop
0        Brazil   Brasilia    10  100
1        Russia     Moscow    20  300
2         India  New Delhi    30  500
3         China    Beijing    40  200
4  South Africa   Pretoria     5   50

In [70]:
df = df.rename(columns = {"Co": "Country", "Cap": "Capital"})

In [71]:
df

Country    Capital  Area  Pop
0        Brazil   Brasilia    10  100
1        Russia     Moscow    20  300
2         India  New Delhi    30  500
3         China    Beijing    40  200
4  South Africa   Pretoria     5   50

In [72]:
df.rename(index = {4: "SA", 3: "Ch"})

Country    Capital  Area  Pop
0         Brazil   Brasilia    10  100
1         Russia     Moscow    20  300
2          India  New Delhi    30  500
Ch         China    Beijing    40  200
SA  South Africa   Pretoria     5   50

In [73]:
df.rename(mapper = str.upper, axis = "columns", inplace = True)

In [74]:
df.set_index("COUNTRY", inplace = True)

In [75]:
df.rename(mapper = str.lower, axis = "rows", inplace = True)

In [76]:
df

CAPITAL  AREA  POP
COUNTRY                           
brazil         Brasilia    10  100
russia           Moscow    20  300
india         New Delhi    30  500
china           Beijing    40  200
south africa   Pretoria     5   50

In [77]:
df.index

Index(['brazil', 'russia', 'india', 'china', 'south africa'], dtype='object', name='COUNTRY')

In [78]:
df["CAPITAL"]

COUNTRY
brazil           Brasilia
russia             Moscow
india           New Delhi
china             Beijing
south africa     Pretoria
Name: CAPITAL, dtype: object

#### How to read the files from external data sources

In [79]:
pwd()

'/home/ubuntu/7thOct'

In [80]:
#workspace or working directory

In [81]:
import os

In [82]:
os.getcwd()

'/home/ubuntu/7thOct'

In [83]:
movies=pd.read_csv("movies.csv", na_values = [1, "?"])

In [84]:
movies.head()

movieId                               title  \
0      NaN                    Toy Story (1995)   
1      2.0                      Jumanji (1995)   
2      3.0             Grumpier Old Men (1995)   
3      4.0            Waiting to Exhale (1995)   
4      5.0  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

In [85]:
#to import the file from a different path (different from working directory)
ap = pd.read_csv("/home/ubuntu/24-Aug/AirPassengers.csv")

In [86]:
ap.head()

Month  #Passengers
0  1949-01          112
1  1949-02          118
2  1949-03          132
3  1949-04          129
4  1949-05          121

In [87]:
t = pd.read_excel("titanic_modified_NA.xlsx")

In [88]:
t.head(20)
#Age row number 6 - blank value read as nan value
#blank values in cabin also read as na

PassengerId Survived  Pclass  \
0             1        0       3   
1             2        1       1   
2             3        1       3   
3             4        1       1   
4             5        0       3   
5             6        0       3   
6             7        0       1   
7             8        0       3   
8             9        1       3   
9            10        1       2   
10           11        1       3   
11           12        1       1   
12           13        0       3   
13           14        0       3   
14           15        0       3   
15           16        1       2   
16           17        0       3   
17           18        ?       2   
18           19        ?       3   
19           20        ?       3   

                                                 Name     Sex   Age  SibSp  \
0                             Braund, Mr. Owen Harris    male  22.0      1   
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                              Heikkinen, Miss. Laina  female  26.0      0   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                            Allen, Mr. William Henry    male  35.0      0   
5                                    Moran, Mr. James    male   NaN      0   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
7                      Palsson, Master. Gosta Leonard    male   2.0      3   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                 Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                    Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   
12                     Saundercock, Mr. William Henry    male  20.0      0   
13                        Andersson, Mr. Anders Johan    male  39.0      1   
14               Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                   Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                               Rice, Master. Eugene    male   2.0      4   
17                       Williams, Mr. Charles Eugene    male   NaN      0   
18  Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                            Masselmani, Mrs. Fatima  female   NaN      0   

    Parch            Ticket     Fare Cabin Embarked  
0       0         A/5 21171   7.2500   NaN        S  
1       0          PC 17599  71.2833   C85        C  
2       0  STON/O2. 3101282   7.9250   NaN        S  
3       0            113803  53.1000  C123        S  
4       0            373450   8.0500   NaN        T  
5       0            330877   8.4583   NaN        Q  
6       0             17463  51.8625   E46        S  
7       1            349909  21.0750   NaN        S  
8       2            347742  11.1333   NaN        S  
9       0            237736  30.0708   NaN        C  
10      1           PP 9549  16.7000    G6        T  
11      0            113783  26.5500  C103        S  
12      0         A/5. 2151   8.0500   NaN        S  
13      5            347082  31.2750   NaN        S  
14      0            350406   7.8542   NaN        S  
15      0            248706  16.0000   NaN        T  
16      1            382652  29.1250   NaN        Q  
17      0            244373  13.0000   NaN        S  
18      0            345763  18.0000   NaN        S  
19      0              2649   7.2250   NaN        T

In [89]:
t = pd.read_excel("titanic_modified_NA.xlsx", na_values = {"Survived": "?", "Embarked": "T"})

In [90]:
t[15:20]
#blank - is the default nan
#T in embarked col is being read as nan
#? in Survived is being read as nan

PassengerId  Survived  Pclass  \
15           16       1.0       2   
16           17       0.0       3   
17           18       NaN       2   
18           19       NaN       3   
19           20       NaN       3   

                                                 Name     Sex   Age  SibSp  \
15                   Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                               Rice, Master. Eugene    male   2.0      4   
17                       Williams, Mr. Charles Eugene    male   NaN      0   
18  Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                            Masselmani, Mrs. Fatima  female   NaN      0   

    Parch  Ticket    Fare Cabin Embarked  
15      0  248706  16.000   NaN      NaN  
16      1  382652  29.125   NaN        Q  
17      0  244373  13.000   NaN        S  
18      0  345763  18.000   NaN        S  
19      0    2649   7.225   NaN      NaN

In [91]:
import warnings
warnings.filterwarnings(action = "ignore")

In [92]:
mov = pd.read_csv("movies.dat", sep = "::", header = None,names = ["MovieID", "MovieName", "Genres"])

In [93]:
mov.head()

MovieID                           MovieName                        Genres
0        1                    Toy Story (1995)   Animation|Children's|Comedy
1        2                      Jumanji (1995)  Adventure|Children's|Fantasy
2        3             Grumpier Old Men (1995)                Comedy|Romance
3        4            Waiting to Exhale (1995)                  Comedy|Drama
4        5  Father of the Bride Part II (1995)                        Comedy

In [94]:
mov = pd.read_table("movies.dat", sep = "::", header = None,names = ["MovieID", "MovieName", "Genres"])

In [95]:
mov.head()

MovieID                           MovieName                        Genres
0        1                    Toy Story (1995)   Animation|Children's|Comedy
1        2                      Jumanji (1995)  Adventure|Children's|Fantasy
2        3             Grumpier Old Men (1995)                Comedy|Romance
3        4            Waiting to Exhale (1995)                  Comedy|Drama
4        5  Father of the Bride Part II (1995)                        Comedy

In [96]:
m = pd.read_csv("data/matches.csv")
#data folder is there in the working directory from there we have to read matches.csv

---

In [97]:
from sklearn import datasets

In [98]:
bos = datasets.load_boston()

In [99]:
type(bos)

sklearn.utils.Bunch

In [100]:
bos

{'data': array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
         4.9800e+00],
        [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
         9.1400e+00],
        [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
         4.0300e+00],
        ...,
        [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
         5.6400e+00],
        [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
         6.4800e+00],
        [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
         7.8800e+00]]),
 'target': array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9, 15. ,
        18.9, 21.7, 20.4, 18.2, 19.9, 23.1, 17.5, 20.2, 18.2, 13.6, 19.6,
        15.2, 14.5, 15.6, 13.9, 16.6, 14.8, 18.4, 21. , 12.7, 14.5, 13.2,
        13.1, 13.5, 18.9, 20. , 21. , 24.7, 30.8, 34.9, 26.6, 25.3, 24.7,
        21.2, 19.3, 20. , 16.6, 14.4, 19.4, 19.7, 20.5, 25. , 23.4, 18.9,
        35.4, 24.7, 3

In [101]:
bos.data.shape

(506, 13)

In [102]:
bos.target.shape

(506,)

In [103]:
boston=np.c_[bos.target,bos.data]

In [104]:
boston.shape

(506, 14)

In [105]:
bos.feature_names.shape

(13,)

In [106]:
columns = ["target"] + list(bos.feature_names)

In [107]:
bs = pd.DataFrame(boston, columns=columns)

In [108]:
bs.head()

target     CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    24.0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1    21.6  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2    34.7  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3    33.4  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4    36.2  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [109]:
ir = datasets.load_iris()

In [110]:
#m for module,
#f for function
#i for instance

In [111]:
bs.dtypes

target     float64
CRIM       float64
ZN         float64
INDUS      float64
CHAS       float64
NOX        float64
RM         float64
AGE        float64
DIS        float64
RAD        float64
TAX        float64
PTRATIO    float64
B          float64
LSTAT      float64
dtype: object

In [112]:
t.dtypes

PassengerId      int64
Survived       float64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [113]:
t.head() #top 5 rows

PassengerId  Survived  Pclass  \
0            1       0.0       3   
1            2       1.0       1   
2            3       1.0       3   
3            4       1.0       1   
4            5       0.0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN      NaN

In [114]:
t.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       887 non-null float64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       884 non-null object
dtypes: float64(3), int64(4), object(5)
memory usage: 83.6+ KB


In [115]:
#so this tells us which columns have got how many null values and their data types

In [116]:
t.describe(include = "all")

PassengerId    Survived      Pclass               Name   Sex  \
count    891.000000  887.000000  891.000000                891   891   
unique          NaN         NaN         NaN                891     2   
top             NaN         NaN         NaN  Lulic, Mr. Nikola  male   
freq            NaN         NaN         NaN                  1   577   
mean     446.000000    0.383315    2.308642                NaN   NaN   
std      257.353842    0.486468    0.836071                NaN   NaN   
min        1.000000    0.000000    1.000000                NaN   NaN   
25%      223.500000    0.000000    2.000000                NaN   NaN   
50%      446.000000    0.000000    3.000000                NaN   NaN   
75%      668.500000    1.000000    3.000000                NaN   NaN   
max      891.000000    1.000000    3.000000                NaN   NaN   

               Age       SibSp       Parch  Ticket        Fare Cabin Embarked  
count   714.000000  891.000000  891.000000   891.0  891.000000   204      884  
unique         NaN         NaN         NaN   681.0         NaN   147        3  
top            NaN         NaN         NaN  1601.0         NaN    G6        S  
freq           NaN         NaN         NaN     7.0         NaN     4      640  
mean     29.699118    0.523008    0.381594     NaN   32.204208   NaN      NaN  
std      14.526497    1.102743    0.806057     NaN   49.693429   NaN      NaN  
min       0.420000    0.000000    0.000000     NaN    0.000000   NaN      NaN  
25%      20.125000    0.000000    0.000000     NaN    7.910400   NaN      NaN  
50%      28.000000    0.000000    0.000000     NaN   14.454200   NaN      NaN  
75%      38.000000    1.000000    0.000000     NaN   31.000000   NaN      NaN  
max      80.000000    8.000000    6.000000     NaN  512.329200   NaN      NaN

25%ile or Quartile 1 value: 25% of the passengers are having age < 20.12
99.9%ile - that only .1% of students have scored more than my score

In [117]:
t.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887       0.0       2                     Montvila, Rev. Juozas   
887          888       1.0       1              Graham, Miss. Margaret Edith   
888          889       0.0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890       1.0       1                     Behr, Mr. Karl Howell   
890          891       0.0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

In [118]:
pd.value_counts(t["Survived"])

0.0    547
1.0    340
Name: Survived, dtype: int64

In [119]:
#t["Name"].value_counts()

In [120]:
t.Sex.value_counts()

male      577
female    314
Name: Sex, dtype: int64

----

#### Subsetting and slicing of data-frames

In [121]:
t[0]

KeyError: 0

In [122]:
t[:1]

PassengerId  Survived  Pclass                     Name   Sex   Age  SibSp  \
0            1       0.0       3  Braund, Mr. Owen Harris  male  22.0      1   

   Parch     Ticket  Fare Cabin Embarked  
0      0  A/5 21171  7.25   NaN        S

In [123]:
t[:4]

PassengerId  Survived  Pclass  \
0            1       0.0       3   
1            2       1.0       1   
2            3       1.0       3   
3            4       1.0       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S

In [124]:
t[10:20:2]

PassengerId  Survived  Pclass  \
10           11       1.0       3   
12           13       0.0       3   
14           15       0.0       3   
16           17       0.0       3   
18           19       NaN       3   

                                                 Name     Sex   Age  SibSp  \
10                    Sandstrom, Miss. Marguerite Rut  female   4.0      1   
12                     Saundercock, Mr. William Henry    male  20.0      0   
14               Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
16                               Rice, Master. Eugene    male   2.0      4   
18  Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   

    Parch     Ticket     Fare Cabin Embarked  
10      1    PP 9549  16.7000    G6      NaN  
12      0  A/5. 2151   8.0500   NaN        S  
14      0     350406   7.8542   NaN        S  
16      1     382652  29.1250   NaN        Q  
18      0     345763  18.0000   NaN        S

In [125]:
df[:3]

CAPITAL  AREA  POP
COUNTRY                      
brazil    Brasilia    10  100
russia      Moscow    20  300
india    New Delhi    30  500

In [126]:
cap = df["CAPITAL"]

In [127]:
type(cap)

pandas.core.series.Series

In [128]:
df[["CAPITAL"]] #returns a df

CAPITAL
COUNTRY                
brazil         Brasilia
russia           Moscow
india         New Delhi
china           Beijing
south africa   Pretoria

In [129]:
df[["CAPITAL","AREA","POP"]]

CAPITAL  AREA  POP
COUNTRY                           
brazil         Brasilia    10  100
russia           Moscow    20  300
india         New Delhi    30  500
china           Beijing    40  200
south africa   Pretoria     5   50

In [130]:
df.CAPITAL=="New Delhi"

COUNTRY
brazil          False
russia          False
india            True
china           False
south africa    False
Name: CAPITAL, dtype: bool

In [131]:
df[df.CAPITAL=="New Delhi"]

CAPITAL  AREA  POP
COUNTRY                      
india    New Delhi    30  500

In [132]:
t[:2]

PassengerId  Survived  Pclass  \
0            1       0.0       3   
1            2       1.0       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  
0      0  A/5 21171   7.2500   NaN        S  
1      0   PC 17599  71.2833   C85        C

In [133]:
t[np.logical_and(t.Survived==0, t.Age<22)].head()

PassengerId  Survived  Pclass                                  Name  \
7             8       0.0       3        Palsson, Master. Gosta Leonard   
12           13       0.0       3        Saundercock, Mr. William Henry   
14           15       0.0       3  Vestrom, Miss. Hulda Amanda Adolfina   
16           17       0.0       3                  Rice, Master. Eugene   
24           25       0.0       3         Palsson, Miss. Torborg Danira   

       Sex   Age  SibSp  Parch     Ticket     Fare Cabin Embarked  
7     male   2.0      3      1     349909  21.0750   NaN        S  
12    male  20.0      0      0  A/5. 2151   8.0500   NaN        S  
14  female  14.0      0      0     350406   7.8542   NaN        S  
16    male   2.0      4      1     382652  29.1250   NaN        Q  
24  female   8.0      3      1     349909  21.0750   NaN        S

#### iloc - integer location

In [134]:
t.iloc[0,0] #0th row number and 0th column number

1

In [135]:
t.iloc[1,0]

2

In [136]:
t.iloc[10,4]

'female'

In [137]:
t.iloc[:10:2, ::-1]

Embarked Cabin     Fare            Ticket  Parch  SibSp   Age     Sex  \
0        S   NaN   7.2500         A/5 21171      0      1  22.0    male   
2        S   NaN   7.9250  STON/O2. 3101282      0      0  26.0  female   
4      NaN   NaN   8.0500            373450      0      0  35.0    male   
6        S   E46  51.8625             17463      0      0  54.0    male   
8        S   NaN  11.1333            347742      2      0  27.0  female   

                                                Name  Pclass  Survived  \
0                            Braund, Mr. Owen Harris       3       0.0   
2                             Heikkinen, Miss. Laina       3       1.0   
4                           Allen, Mr. William Henry       3       0.0   
6                            McCarthy, Mr. Timothy J       1       0.0   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)       3       1.0   

   PassengerId  
0            1  
2            3  
4            5  
6            7  
8            9

In [138]:
t.iloc[0] #0th row number in the series format

PassengerId                          1
Survived                             0
Pclass                               3
Name           Braund, Mr. Owen Harris
Sex                               male
Age                                 22
SibSp                                1
Parch                                0
Ticket                       A/5 21171
Fare                              7.25
Cabin                              NaN
Embarked                             S
Name: 0, dtype: object

In [139]:
t.iloc[[0]] #this gives a df

PassengerId  Survived  Pclass                     Name   Sex   Age  SibSp  \
0            1       0.0       3  Braund, Mr. Owen Harris  male  22.0      1   

   Parch     Ticket  Fare Cabin Embarked  
0      0  A/5 21171  7.25   NaN        S

In [140]:
t.iloc[:, [2]].head() #this returns a df

Pclass
0       3
1       1
2       3
3       1
4       3

In [141]:
t.iloc[[3,0,5], [0,1,2]]

PassengerId  Survived  Pclass
3            4       1.0       1
0            1       0.0       3
5            6       0.0       3

In [142]:
t.iloc[[10,4]]

PassengerId  Survived  Pclass                             Name     Sex  \
10           11       1.0       3  Sandstrom, Miss. Marguerite Rut  female   
4             5       0.0       3         Allen, Mr. William Henry    male   

     Age  SibSp  Parch   Ticket   Fare Cabin Embarked  
10   4.0      1      1  PP 9549  16.70    G6      NaN  
4   35.0      0      0   373450   8.05   NaN      NaN

In [143]:
t.iloc[[10,4], :]

PassengerId  Survived  Pclass                             Name     Sex  \
10           11       1.0       3  Sandstrom, Miss. Marguerite Rut  female   
4             5       0.0       3         Allen, Mr. William Henry    male   

     Age  SibSp  Parch   Ticket   Fare Cabin Embarked  
10   4.0      1      1  PP 9549  16.70    G6      NaN  
4   35.0      0      0   373450   8.05   NaN      NaN

In [144]:
df.iloc[[3,4], [-1]]

POP
COUNTRY          
china         200
south africa   50

#### loc command
- specify by row index and the column name

In [145]:
t.head(2)

PassengerId  Survived  Pclass  \
0            1       0.0       3   
1            2       1.0       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  
0      0  A/5 21171   7.2500   NaN        S  
1      0   PC 17599  71.2833   C85        C

In [146]:
t.loc[:, ["Survived", "Name", "Age", "Sex"]].head()

Survived                                               Name   Age     Sex
0       0.0                            Braund, Mr. Owen Harris  22.0    male
1       1.0  Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0  female
2       1.0                             Heikkinen, Miss. Laina  26.0  female
3       1.0       Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0  female
4       0.0                           Allen, Mr. William Henry  35.0    male

In [147]:
t.loc[:5, ["Survived"]]

Survived
0       0.0
1       1.0
2       1.0
3       1.0
4       0.0
5       0.0

In [148]:
df.loc["brazil"]

CAPITAL    Brasilia
AREA             10
POP             100
Name: brazil, dtype: object

In [149]:
df.loc[["brazil"]]

CAPITAL  AREA  POP
COUNTRY                     
brazil   Brasilia    10  100

In [150]:
df.loc[["china"]]

CAPITAL  AREA  POP
COUNTRY                    
china    Beijing    40  200

In [151]:
df

CAPITAL  AREA  POP
COUNTRY                           
brazil         Brasilia    10  100
russia           Moscow    20  300
india         New Delhi    30  500
china           Beijing    40  200
south africa   Pretoria     5   50

In [152]:
df.loc["brazil": "china", "CAPITAL": "AREA"]

CAPITAL  AREA
COUNTRY                 
brazil    Brasilia    10
russia      Moscow    20
india    New Delhi    30
china      Beijing    40

In [153]:
t.loc[10:14, "Pclass"] #returns series

10    3
11    1
12    3
13    3
14    3
Name: Pclass, dtype: int64

In [154]:
t.loc[10:14, ["Pclass"]]

Pclass
10       3
11       1
12       3
13       3
14       3

In [155]:
t.head()

PassengerId  Survived  Pclass  \
0            1       0.0       3   
1            2       1.0       1   
2            3       1.0       3   
3            4       1.0       1   
4            5       0.0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN      NaN

In [156]:
t.loc[t.Sex=="male", ["Parch"]].head()

Parch
0      0
4      0
5      0
6      0
7      1

In [157]:
df

CAPITAL  AREA  POP
COUNTRY                           
brazil         Brasilia    10  100
russia           Moscow    20  300
india         New Delhi    30  500
china           Beijing    40  200
south africa   Pretoria     5   50

In [158]:
df.reset_index(inplace = True)

In [159]:
df

COUNTRY    CAPITAL  AREA  POP
0        brazil   Brasilia    10  100
1        russia     Moscow    20  300
2         india  New Delhi    30  500
3         china    Beijing    40  200
4  south africa   Pretoria     5   50

In [160]:
countries = ["USA", "China", "Russia", "Pakistan", "Afghanistan"]

In [161]:
df.COUNTRY = df.COUNTRY.str.capitalize()

In [162]:
df

COUNTRY    CAPITAL  AREA  POP
0        Brazil   Brasilia    10  100
1        Russia     Moscow    20  300
2         India  New Delhi    30  500
3         China    Beijing    40  200
4  South africa   Pretoria     5   50

In [163]:
"Brazil" in countries

False

In [164]:
"Russia" not in countries

False

In [165]:
countries

['USA', 'China', 'Russia', 'Pakistan', 'Afghanistan']

In [166]:
df.COUNTRY

0          Brazil
1          Russia
2           India
3           China
4    South africa
Name: COUNTRY, dtype: object

In [167]:
lst = []
for x in df.COUNTRY:
    if(x in countries):
       lst.append(x) 

In [168]:
lst

['Russia', 'China']

In [169]:
df.loc[df.COUNTRY.isin(countries), ["COUNTRY"]]

COUNTRY
1  Russia
3   China

In [170]:
df.loc[:, "POP"::-1]

POP  AREA    CAPITAL       COUNTRY
0  100    10   Brasilia        Brazil
1  300    20     Moscow        Russia
2  500    30  New Delhi         India
3  200    40    Beijing         China
4   50     5   Pretoria  South africa

-----

#### Statistical functions using axis

In [171]:
df.mean()

AREA     21.0
POP     230.0
dtype: float64

In [172]:
#mean median sum min max argmin argmax quantile
#axis = 0 you are applyin the reduce operation row-wise - which means we are getting column mean

In [173]:
df.mean(axis = 0)

AREA     21.0
POP     230.0
dtype: float64

In [174]:
t.Age.min()

0.42

In [175]:
t['Age'].median()

28.0

In [176]:
t['Age'].quantile(.99)

65.87

In [177]:
np.percentile(t.Age, 99) #giving me nan values because there are missing values in my data

nan

In [178]:
np.percentile(df.POP, 99)

492.0

In [179]:
t.var(axis = 0)

PassengerId    66231.000000
Survived           0.236651
Pclass             0.699015
Age              211.019125
SibSp              1.216043
Parch              0.649728
Fare            2469.436846
dtype: float64

In [180]:
t["Agecappedat30"] = np.where(t.Age>30, 30, t.Age)

In [181]:
t[:2]

PassengerId  Survived  Pclass  \
0            1       0.0       3   
1            2       1.0       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  Agecappedat30  
0      0  A/5 21171   7.2500   NaN        S           22.0  
1      0   PC 17599  71.2833   C85        C           30.0

In [182]:
np.where(t.Age>30)# to retreive the row numbers where the cond is being satisfied

(array([  1,   3,   4,   6,  11,  13,  15,  18,  20,  21,  25,  30,  33,
         35,  40,  52,  54,  61,  62,  70,  74,  85,  92,  94,  96,  98,
         99, 103, 104, 108, 110, 116, 122, 123, 124, 129, 130, 132, 137,
        148, 149, 150, 152, 153, 155, 160, 161, 167, 170, 174, 177, 179,
        187, 188, 189, 190, 194, 195, 197, 202, 203, 206, 209, 211, 215,
        217, 218, 222, 224, 230, 232, 236, 239, 245, 248, 249, 252, 254,
        258, 259, 262, 263, 265, 268, 269, 272, 273, 275, 276, 279, 280,
        285, 288, 292, 299, 314, 317, 318, 319, 325, 326, 327, 328, 331,
        332, 337, 338, 339, 344, 346, 349, 357, 360, 362, 363, 366, 380,
        382, 383, 387, 390, 396, 397, 400, 405, 406, 412, 414, 416, 429,
        432, 434, 438, 439, 440, 447, 449, 450, 453, 456, 458, 460, 461,
        462, 463, 465, 467, 471, 472, 476, 482, 483, 486, 487, 492, 493,
        496, 503, 506, 512, 513, 515, 516, 518, 519, 523, 525, 526, 528,
        536, 540, 543, 544, 545, 548, 555, 556, 558

----

---

# Class 8_Part 2

#### Concatenations

In [183]:
df

COUNTRY    CAPITAL  AREA  POP
0        Brazil   Brasilia    10  100
1        Russia     Moscow    20  300
2         India  New Delhi    30  500
3         China    Beijing    40  200
4  South africa   Pretoria     5   50

In [184]:
row = {"COUNTRY": ["Brazil"], "CAPITAL": ["Brasilia"], "AR": [100], "POP": [10]}

In [185]:
dummy = pd.DataFrame(row)
dummy

COUNTRY   CAPITAL   AR  POP
0  Brazil  Brasilia  100   10

In [186]:
df=df.append(dummy)
df

AR  AREA    CAPITAL       COUNTRY  POP
0    NaN  10.0   Brasilia        Brazil  100
1    NaN  20.0     Moscow        Russia  300
2    NaN  30.0  New Delhi         India  500
3    NaN  40.0    Beijing         China  200
4    NaN   5.0   Pretoria  South africa   50
0  100.0   NaN   Brasilia        Brazil   10

In [187]:
df.reset_index(inplace = True)

In [188]:
df = df.drop("index", axis =1)

In [189]:
df

AR  AREA    CAPITAL       COUNTRY  POP
0    NaN  10.0   Brasilia        Brazil  100
1    NaN  20.0     Moscow        Russia  300
2    NaN  30.0  New Delhi         India  500
3    NaN  40.0    Beijing         China  200
4    NaN   5.0   Pretoria  South africa   50
5  100.0   NaN   Brasilia        Brazil   10

In [190]:
df=pd.concat([df, dummy], axis = 0)

In [191]:
df

AR  AREA    CAPITAL       COUNTRY  POP
0    NaN  10.0   Brasilia        Brazil  100
1    NaN  20.0     Moscow        Russia  300
2    NaN  30.0  New Delhi         India  500
3    NaN  40.0    Beijing         China  200
4    NaN   5.0   Pretoria  South africa   50
5  100.0   NaN   Brasilia        Brazil   10
0  100.0   NaN   Brasilia        Brazil   10

----

#### Inner and Outer joins

In [192]:
left = {"id": [1,2,3,4,5], 
       "Name": ["Alex", "Amy", "Allen", "Alice", "Ayoung"],
       "subject_id": ["sub1", "sub2", "sub4", "sub3", "sub5"]}

In [193]:
right = {"id": [1,6,3, 4,5], "Height": [120, 110, 115, 119, 130], "Subject_ID": ["sub1", "sub2", "sub3", "sub4","sub5"]}

In [194]:
left = pd.DataFrame(left)

In [195]:
right = pd.DataFrame(right)

In [196]:
print(left)
print("\n")
print(right)

   id    Name subject_id
0   1    Alex       sub1
1   2     Amy       sub2
2   3   Allen       sub4
3   4   Alice       sub3
4   5  Ayoung       sub5


   id  Height Subject_ID
0   1     120       sub1
1   6     110       sub2
2   3     115       sub3
3   4     119       sub4
4   5     130       sub5


In [197]:
inner_join = pd.merge(left = left, right=right, on = "id")

In [198]:
inner_join

id    Name subject_id  Height Subject_ID
0   1    Alex       sub1     120       sub1
1   3   Allen       sub4     115       sub3
2   4   Alice       sub3     119       sub4
3   5  Ayoung       sub5     130       sub5

In [199]:
left_outer_join = pd.merge(left = left, right=right, on = "id", how = "left")

In [200]:
print(left)
print(right)

   id    Name subject_id
0   1    Alex       sub1
1   2     Amy       sub2
2   3   Allen       sub4
3   4   Alice       sub3
4   5  Ayoung       sub5
   id  Height Subject_ID
0   1     120       sub1
1   6     110       sub2
2   3     115       sub3
3   4     119       sub4
4   5     130       sub5


In [201]:
left_outer_join

id    Name subject_id  Height Subject_ID
0   1    Alex       sub1   120.0       sub1
1   2     Amy       sub2     NaN        NaN
2   3   Allen       sub4   115.0       sub3
3   4   Alice       sub3   119.0       sub4
4   5  Ayoung       sub5   130.0       sub5

-----

In [202]:
#!pip install pandasql

In [203]:
t.head()

PassengerId  Survived  Pclass  \
0            1       0.0       3   
1            2       1.0       1   
2            3       1.0       3   
3            4       1.0       1   
4            5       0.0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Agecappedat30  
0      0         A/5 21171   7.2500   NaN        S           22.0  
1      0          PC 17599  71.2833   C85        C           30.0  
2      0  STON/O2. 3101282   7.9250   NaN        S           26.0  
3      0            113803  53.1000  C123        S           30.0  
4      0            373450   8.0500   NaN      NaN           30.0

In [204]:
str1 = "select Name, Age from t where Age>35 or SibSp<1"

In [205]:
!pip install pandasql

ERROR: Could not install packages due to an EnvironmentError: [Errno 30] Read-only file system: '/opt/anaconda3/lib/python3.7/site-packages/pandasql-0.7.3.dist-info'



In [206]:
import pandasql as pdsql
#Pandasql will be used for data manipulation using sql queries

ModuleNotFoundError: No module named 'pandasql'

In [207]:
var= pdsql.sqldf(str1)

NameError: name 'pdsql' is not defined

----

### How to apply a string function on the entire string column

In [218]:
t.Name = t.Name.str.upper()

In [219]:
t["title"] = t.Name.str.split("[, .]").str[2].str.capitalize()

In [220]:
t.head()

PassengerId  Survived  Pclass  \
0            1       0.0       3   
1            2       1.0       1   
2            3       1.0       3   
3            4       1.0       1   
4            5       0.0       3   

                                                Name     Sex   Age  SibSp  \
0                            BRAUND, MR. OWEN HARRIS    male  22.0      1   
1  CUMINGS, MRS. JOHN BRADLEY (FLORENCE BRIGGS TH...  female  38.0      1   
2                             HEIKKINEN, MISS. LAINA  female  26.0      0   
3       FUTRELLE, MRS. JACQUES HEATH (LILY MAY PEEL)  female  35.0      1   
4                           ALLEN, MR. WILLIAM HENRY    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Agecappedat30  \
0      0         A/5 21171   7.2500   NaN        S           22.0   
1      0          PC 17599  71.2833   C85        C           30.0   
2      0  STON/O2. 3101282   7.9250   NaN        S           26.0   
3      0            113803  53.1000  C123        S           30.0   
4      0            373450   8.0500   NaN      NaN           30.0   

   family_size  f_size title  
0            2     2.0    Mr  
1            2     2.0   Mrs  
2            1     1.0  Miss  
3            2     2.0   Mrs  
4            1     1.0    Mr

#### to iterate over each row of the dataframe

In [221]:
len(df)

7

In [222]:
for row in range(len(df)): #len df gives the number of rows
    print(df.iloc[row],"\n")

AR              NaN
AREA             10
CAPITAL    Brasilia
POP             100
Name: Brazil, dtype: object 

AR            NaN
AREA           20
CAPITAL    Moscow
POP           300
Name: Russia, dtype: object 

AR               NaN
AREA              30
CAPITAL    New Delhi
POP              500
Name: India, dtype: object 

AR             NaN
AREA            40
CAPITAL    Beijing
POP            200
Name: China, dtype: object 

AR              NaN
AREA              5
CAPITAL    Pretoria
POP              50
Name: South africa, dtype: object 

AR              100
AREA            NaN
CAPITAL    Brasilia
POP              10
Name: Brazil, dtype: object 

AR              100
AREA            NaN
CAPITAL    Brasilia
POP              10
Name: Brazil, dtype: object 



In [223]:
df.set_index("COUNTRY", inplace = True)

KeyError: 'COUNTRY'

In [ ]:
df

In [224]:
for lab, row in df.iterrows(): #lab will get the row index and row will get the row
    print(lab)
    print(row)
    print(df.loc[lab,:])
    print()

Brazil
AR              NaN
AREA             10
CAPITAL    Brasilia
POP             100
Name: Brazil, dtype: object
            AR  AREA   CAPITAL  POP
COUNTRY                            
Brazil     NaN  10.0  Brasilia  100
Brazil   100.0   NaN  Brasilia   10
Brazil   100.0   NaN  Brasilia   10

Russia
AR            NaN
AREA           20
CAPITAL    Moscow
POP           300
Name: Russia, dtype: object
AR            NaN
AREA           20
CAPITAL    Moscow
POP           300
Name: Russia, dtype: object

India
AR               NaN
AREA              30
CAPITAL    New Delhi
POP              500
Name: India, dtype: object
AR               NaN
AREA              30
CAPITAL    New Delhi
POP              500
Name: India, dtype: object

China
AR             NaN
AREA            40
CAPITAL    Beijing
POP            200
Name: China, dtype: object
AR             NaN
AREA            40
CAPITAL    Beijing
POP            200
Name: China, dtype: object

South africa
AR              NaN
AREA              5
C

In [225]:
t[:2]

PassengerId  Survived  Pclass  \
0            1       0.0       3   
1            2       1.0       1   

                                                Name     Sex   Age  SibSp  \
0                            BRAUND, MR. OWEN HARRIS    male  22.0      1   
1  CUMINGS, MRS. JOHN BRADLEY (FLORENCE BRIGGS TH...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  Agecappedat30  family_size  \
0      0  A/5 21171   7.2500   NaN        S           22.0            2   
1      0   PC 17599  71.2833   C85        C           30.0            2   

   f_size title  
0     2.0    Mr  
1     2.0   Mrs

In [226]:
t["family_size"] = t.SibSp + t.Parch + 1

In [227]:
for idx, row in t.iterrows():
    t.loc[idx, "f_size"] = t.loc[idx, "SibSp"] + t.loc[idx, "Parch"] + 1

In [228]:
t["f_size"].head()

0    2.0
1    2.0
2    1.0
3    2.0
4    1.0
Name: f_size, dtype: float64

In [229]:
t.head()

PassengerId  Survived  Pclass  \
0            1       0.0       3   
1            2       1.0       1   
2            3       1.0       3   
3            4       1.0       1   
4            5       0.0       3   

                                                Name     Sex   Age  SibSp  \
0                            BRAUND, MR. OWEN HARRIS    male  22.0      1   
1  CUMINGS, MRS. JOHN BRADLEY (FLORENCE BRIGGS TH...  female  38.0      1   
2                             HEIKKINEN, MISS. LAINA  female  26.0      0   
3       FUTRELLE, MRS. JACQUES HEATH (LILY MAY PEEL)  female  35.0      1   
4                           ALLEN, MR. WILLIAM HENRY    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Agecappedat30  \
0      0         A/5 21171   7.2500   NaN        S           22.0   
1      0          PC 17599  71.2833   C85        C           30.0   
2      0  STON/O2. 3101282   7.9250   NaN        S           26.0   
3      0            113803  53.1000  C123        S           30.0   
4      0            373450   8.0500   NaN      NaN           30.0   

   family_size  f_size title  
0            2     2.0    Mr  
1            2     2.0   Mrs  
2            1     1.0  Miss  
3            2     2.0   Mrs  
4            1     1.0    Mr

In [230]:
t["title_gender"] = t["title"] + " is " + t["Sex"]

In [231]:
t["title_gender"].value_counts()

Mr is male              502
Miss is female          179
Mrs is female           121
Master is male           40
 is male                 13
Dr is male                6
 is female                6
Rev is male               6
Mlle is female            2
Col is male               2
Major is male             2
Don is male               1
Hoef is male              1
Castellana is female      1
Jonkheer is male          1
The is female             1
Dr is female              1
Manent is male            1
Capt is male              1
Ms is female              1
Castellana is male        1
Mme is female             1
More is female            1
Name: title_gender, dtype: int64

---

### Apply function - used to apply some function on each component of the series

In [232]:
x = pd.Series([29,12, 13], index = ["London", "NewYork", "California"])

In [233]:
x

London        29
NewYork       12
California    13
dtype: int64

In [234]:
x.apply(np.log)

London        3.367296
NewYork       2.484907
California    2.564949
dtype: float64

In [235]:
t["Log_Fare"] = t["Fare"].apply(np.log)

In [236]:
t["Log_Fare"] = t["Fare"].apply(lambda x: np.log(x))

In [237]:
t.Log_Fare.head()

0    1.981001
1    4.266662
2    2.070022
3    3.972177
4    2.085672
Name: Log_Fare, dtype: float64

In [238]:
def add_custom_vales(days_b4_june, *args, **kwargs):
    #*args - any number of arguments can come
    #**kwargs - is a dictionary key value pair of these arguments
    for var in kwargs:
        days_b4_june+= kwargs[var]
    return days_b4_june

In [241]:
add_custom_vales(days_b4_june= 10, june = 10, july = 20, aug = 5, sep = 10, oct = 19)

74

In [240]:
x.apply(add_custom_vales, june=10, july=20, a = 100)

London        159
NewYork       142
California    143
dtype: int64

---

### Map Function

In [242]:
df

AR  AREA    CAPITAL  POP
COUNTRY                                  
Brazil          NaN  10.0   Brasilia  100
Russia          NaN  20.0     Moscow  300
India           NaN  30.0  New Delhi  500
China           NaN  40.0    Beijing  200
South africa    NaN   5.0   Pretoria   50
Brazil        100.0   NaN   Brasilia   10
Brazil        100.0   NaN   Brasilia   10

In [244]:
df = df.drop("AR", axis = 1)

In [245]:
df['CAPITAL'].map(len) 

COUNTRY
Brazil          8
Russia          6
India           9
China           7
South africa    8
Brazil          8
Brazil          8
Name: CAPITAL, dtype: int64

In [246]:
t.head()

PassengerId  Survived  Pclass  \
0            1       0.0       3   
1            2       1.0       1   
2            3       1.0       3   
3            4       1.0       1   
4            5       0.0       3   

                                                Name     Sex   Age  SibSp  \
0                            BRAUND, MR. OWEN HARRIS    male  22.0      1   
1  CUMINGS, MRS. JOHN BRADLEY (FLORENCE BRIGGS TH...  female  38.0      1   
2                             HEIKKINEN, MISS. LAINA  female  26.0      0   
3       FUTRELLE, MRS. JACQUES HEATH (LILY MAY PEEL)  female  35.0      1   
4                           ALLEN, MR. WILLIAM HENRY    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Agecappedat30  \
0      0         A/5 21171   7.2500   NaN        S           22.0   
1      0          PC 17599  71.2833   C85        C           30.0   
2      0  STON/O2. 3101282   7.9250   NaN        S           26.0   
3      0            113803  53.1000  C123        S           30.0   
4      0            373450   8.0500   NaN      NaN           30.0   

   family_size  f_size title    title_gender  Log_Fare  
0            2     2.0    Mr      Mr is male  1.981001  
1            2     2.0   Mrs   Mrs is female  4.266662  
2            1     1.0  Miss  Miss is female  2.070022  
3            2     2.0   Mrs   Mrs is female  3.972177  
4            1     1.0    Mr      Mr is male  2.085672

#### Map can be used to map a dictionary / a lambda function

In [251]:
t["Sex"] = t["Sex"].map({'female': 1, "male":0}) #mapped a dictionary female to 1 and male to 0 

In [253]:
t[:2]

PassengerId  Survived  Pclass  \
0            1       0.0       3   
1            2       1.0       1   

                                                Name  Sex   Age  SibSp  Parch  \
0                            BRAUND, MR. OWEN HARRIS    0  22.0      1      0   
1  CUMINGS, MRS. JOHN BRADLEY (FLORENCE BRIGGS TH...    1  38.0      1      0   

      Ticket     Fare Cabin Embarked  Agecappedat30  family_size  f_size  \
0  A/5 21171   7.2500   NaN        S           22.0            2     2.0   
1   PC 17599  71.2833   C85        C           30.0            2     2.0   

  title   title_gender  Log_Fare  
0    Mr     Mr is male  1.981001  
1   Mrs  Mrs is female  4.266662

In [254]:
t.Name[0]

'BRAUND, MR. OWEN HARRIS'

In [256]:
t["title"] = t.Name.map(lambda x: x.split()[1])

In [257]:
df

AREA    CAPITAL  POP
COUNTRY                           
Brazil        10.0   Brasilia  100
Russia        20.0     Moscow  300
India         30.0  New Delhi  500
China         40.0    Beijing  200
South africa   5.0   Pretoria   50
Brazil         NaN   Brasilia   10
Brazil         NaN   Brasilia   10

In [259]:
df.reset_index(inplace = True)

In [261]:
df["COUNTRY"].replace("Brazil", "Ru")

0              Ru
1          Russia
2           India
3           China
4    South africa
5              Ru
6              Ru
Name: COUNTRY, dtype: object

In [262]:
t[:2]

PassengerId  Survived  Pclass  \
0            1       0.0       3   
1            2       1.0       1   

                                                Name  Sex   Age  SibSp  Parch  \
0                            BRAUND, MR. OWEN HARRIS    0  22.0      1      0   
1  CUMINGS, MRS. JOHN BRADLEY (FLORENCE BRIGGS TH...    1  38.0      1      0   

      Ticket     Fare Cabin Embarked  Agecappedat30  family_size  f_size  \
0  A/5 21171   7.2500   NaN        S           22.0            2     2.0   
1   PC 17599  71.2833   C85        C           30.0            2     2.0   

  title   title_gender  Log_Fare  
0   MR.     Mr is male  1.981001  
1  MRS.  Mrs is female  4.266662

In [264]:
t["Sex"] = t.Sex.replace({0:"male", 1:"female"})

In [265]:
t[:2]

PassengerId  Survived  Pclass  \
0            1       0.0       3   
1            2       1.0       1   

                                                Name     Sex   Age  SibSp  \
0                            BRAUND, MR. OWEN HARRIS    male  22.0      1   
1  CUMINGS, MRS. JOHN BRADLEY (FLORENCE BRIGGS TH...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  Agecappedat30  family_size  \
0      0  A/5 21171   7.2500   NaN        S           22.0            2   
1      0   PC 17599  71.2833   C85        C           30.0            2   

   f_size title   title_gender  Log_Fare  
0     2.0   MR.     Mr is male  1.981001  
1     2.0  MRS.  Mrs is female  4.266662

In [267]:
t.Embarked.replace({np.NaN: "A"})

0      S
1      C
2      S
3      S
4      A
5      Q
6      S
7      S
8      S
9      C
10     A
11     S
12     S
13     S
14     S
15     A
16     Q
17     S
18     S
19     A
20     S
21     S
22     Q
23     S
24     S
25     S
26     C
27     A
28     Q
29     S
      ..
861    S
862    S
863    S
864    S
865    S
866    C
867    S
868    S
869    S
870    S
871    S
872    S
873    S
874    C
875    C
876    S
877    S
878    S
879    C
880    S
881    S
882    S
883    S
884    S
885    Q
886    S
887    S
888    S
889    C
890    Q
Name: Embarked, Length: 891, dtype: object

In [269]:
t.Embarked.map({np.NaN: "A"})

0      NaN
1      NaN
2      NaN
3      NaN
4        A
5      NaN
6      NaN
7      NaN
8      NaN
9      NaN
10       A
11     NaN
12     NaN
13     NaN
14     NaN
15       A
16     NaN
17     NaN
18     NaN
19       A
20     NaN
21     NaN
22     NaN
23     NaN
24     NaN
25     NaN
26     NaN
27       A
28     NaN
29     NaN
      ... 
861    NaN
862    NaN
863    NaN
864    NaN
865    NaN
866    NaN
867    NaN
868    NaN
869    NaN
870    NaN
871    NaN
872    NaN
873    NaN
874    NaN
875    NaN
876    NaN
877    NaN
878    NaN
879    NaN
880    NaN
881    NaN
882    NaN
883    NaN
884    NaN
885    NaN
886    NaN
887    NaN
888    NaN
889    NaN
890    NaN
Name: Embarked, Length: 891, dtype: object

In [271]:
t["Sex"].map({"male": 0})

0      0.0
1      NaN
2      NaN
3      NaN
4      0.0
5      0.0
6      0.0
7      0.0
8      NaN
9      NaN
10     NaN
11     NaN
12     0.0
13     0.0
14     NaN
15     NaN
16     0.0
17     0.0
18     NaN
19     NaN
20     0.0
21     0.0
22     NaN
23     0.0
24     NaN
25     NaN
26     0.0
27     0.0
28     NaN
29     0.0
      ... 
861    0.0
862    NaN
863    NaN
864    0.0
865    NaN
866    NaN
867    0.0
868    0.0
869    0.0
870    0.0
871    NaN
872    0.0
873    0.0
874    NaN
875    NaN
876    0.0
877    0.0
878    0.0
879    NaN
880    NaN
881    0.0
882    NaN
883    0.0
884    0.0
885    NaN
886    0.0
887    NaN
888    NaN
889    0.0
890    0.0
Name: Sex, Length: 891, dtype: float64

In [ ]:
#replace can replace a single category also, where-as map has to provide the mapping for each category of the column

In [275]:
df.COUNTRY.replace("^B.+", "India", regex = True) #replace fn can use reg. expressions for replacement unlike map fn

0           India
1          Russia
2           India
3           China
4    South africa
5           India
6           India
Name: COUNTRY, dtype: object

In [276]:
data = {'raw': ['Arizona 1 2014-12-23       3242.0',
                'Iowa 1 2010-02-23       3453.7',
                'Oregon 0 2014-06-20       2123.0',
                'Maryland 0 2014-03-14       1123.6',
                'Florida 1 2013-01-15       2134.0',
                'Georgia 0 2012-07-14       2345.6']}

In [277]:
data

{'raw': ['Arizona 1 2014-12-23       3242.0',
  'Iowa 1 2010-02-23       3453.7',
  'Oregon 0 2014-06-20       2123.0',
  'Maryland 0 2014-03-14       1123.6',
  'Florida 1 2013-01-15       2134.0',
  'Georgia 0 2012-07-14       2345.6']}

In [278]:
df = pd.DataFrame(data, columns=["raw"])

In [279]:
df

raw
0   Arizona 1 2014-12-23       3242.0
1      Iowa 1 2010-02-23       3453.7
2    Oregon 0 2014-06-20       2123.0
3  Maryland 0 2014-03-14       1123.6
4   Florida 1 2013-01-15       2134.0
5   Georgia 0 2012-07-14       2345.6

In [280]:
a = df.iloc[0,0]

In [281]:
a

'Arizona 1 2014-12-23       3242.0'

In [283]:
import re
b = re.search('(\d)', a)
b.group()

'1'

In [287]:
df["Gender"]=df["raw"].str.extract("(\d)")

In [288]:
df

raw Gender
0   Arizona 1 2014-12-23       3242.0      1
1      Iowa 1 2010-02-23       3453.7      1
2    Oregon 0 2014-06-20       2123.0      0
3  Maryland 0 2014-03-14       1123.6      0
4   Florida 1 2013-01-15       2134.0      1
5   Georgia 0 2012-07-14       2345.6      0

In [289]:
df["date"] = df["raw"].str.extract("(\d\d\d\d-\d\d-\d\d)")

In [290]:
df

raw Gender        date
0   Arizona 1 2014-12-23       3242.0      1  2014-12-23
1      Iowa 1 2010-02-23       3453.7      1  2010-02-23
2    Oregon 0 2014-06-20       2123.0      0  2014-06-20
3  Maryland 0 2014-03-14       1123.6      0  2014-03-14
4   Florida 1 2013-01-15       2134.0      1  2013-01-15
5   Georgia 0 2012-07-14       2345.6      0  2012-07-14

In [292]:
df["State"]=df["raw"].str.extract("([A-Z]\w+)")

In [293]:
df

raw Gender        date     State
0   Arizona 1 2014-12-23       3242.0      1  2014-12-23   Arizona
1      Iowa 1 2010-02-23       3453.7      1  2010-02-23      Iowa
2    Oregon 0 2014-06-20       2123.0      0  2014-06-20    Oregon
3  Maryland 0 2014-03-14       1123.6      0  2014-03-14  Maryland
4   Florida 1 2013-01-15       2134.0      1  2013-01-15   Florida
5   Georgia 0 2012-07-14       2345.6      0  2012-07-14   Georgia

In [297]:
df["Score"] = df["raw"].str.extract("(\d\d\d\d\.\d)")

In [298]:
df

raw Gender        date     State   Score
0   Arizona 1 2014-12-23       3242.0      1  2014-12-23   Arizona  3242.0
1      Iowa 1 2010-02-23       3453.7      1  2010-02-23      Iowa  3453.7
2    Oregon 0 2014-06-20       2123.0      0  2014-06-20    Oregon  2123.0
3  Maryland 0 2014-03-14       1123.6      0  2014-03-14  Maryland  1123.6
4   Florida 1 2013-01-15       2134.0      1  2013-01-15   Florida  2134.0
5   Georgia 0 2012-07-14       2345.6      0  2012-07-14   Georgia  2345.6

In [301]:
t.head(1)

PassengerId  Survived  Pclass                     Name   Sex   Age  SibSp  \
0            1       0.0       3  BRAUND, MR. OWEN HARRIS  male  22.0      1   

   Parch     Ticket  Fare Cabin Embarked  Agecappedat30  family_size  f_size  \
0      0  A/5 21171  7.25   NaN        S           22.0            2     2.0   

  title title_gender  Log_Fare  
0   MR.   Mr is male  1.981001

In [304]:
t["Title"] = t.Name.str.extract("([A-Z]+)\.")